### PR TITLE
Keep the log mean's correction where the mean itself is the correction

### PR DESCRIPTION
**Pull request by choxos**
## What was wrong

`.weighted_log_mean_exp()` shifts by the largest value and subtracts two log sums. Close to a probability of one that correction IS the answer, and the shift discards it.

The cloglog non-event log probability is `-exp(eta)`. At `eta = c(-40, -39)` the two values are `c(-4.25e-18, -1.15e-17)`; the shifted difference is -7.3e-18, `exp()` of it rounds to exactly 1, the two log sums cancel to zero, and the bare maximum is returned:

```
computed log q-bar : -4.24835425529158887095e-18
true     log q-bar : -7.89828921415368749514e-18
```

A relative error of 46%, in a quantity the cloglog link gradient divides by.

## Consequence

| target etas | weights | before | correct |
|---|---|---:|---:|
| (-40, -39) | (1, 1) | 1.8591409142 | 1.0000000000 |
| (-40, -39) | (1, 3) | 2.2887113713 | 1.0000000000 |
| (-40, -39), `X = cbind(1, c(0,1))` | (1, 1) | (1.859, 1.359) | (1, e/(1+e)) |

`1.8591409142 = (1 + e) / 2` exactly, which is the ratio of the true mean to the returned maximum. The index component of the delta-method variance was inflated by about 1.86 while the reported link stayed stable at -39.38, so the estimate looked fine and only its uncertainty was wrong.

## The repair

`log1p(sum(w * expm1(x)) / sum(w))` is the same mean with nothing to cancel: each `expm1(x_i)` keeps full relative accuracy for tiny `x_i`, the terms share a sign so their sum does too, and `log1p` inverts it. Taken only where it is also the safe form, a maximum in `(-1, 0]`, since `expm1` of a large positive value overflows and a mean that underflowed to zero would come back as `log1p(-1)`. Outside that range the shift is the accurate one.

The exact identical-value invariant is now stated outright instead of emerging from the cancellation, since the shares are built as `exp(x - mean)` and must sum to one. That is what the upper-tail replication fix rests on, and `log1p(expm1(x))` need not round-trip to the last bit.

## Measured

Against a 60-digit reference over 122 designs spanning the cloglog tails, linear-predictor inputs and near-zero inputs:

| | relative error > 1e-13 | worst |
|---|---:|---:|
| before | 48 of 122 | 0.998 |
| after | 0 of 122 | 3.6e-16 |

So this was wrong across the whole near-one-probability regime, not only at the reported point.

Failing-first: 10 assertions in the two touched test files fail against the pre-fix source and none do after. Full suite 4713 pass, 0 fail, 0 error.

Upper-tail and identical-profile invariants are asserted in the same test, since a lower-tail repair that reinstated the replication defect would be no repair.

No NEWS entry: `.weighted_log_mean_exp()` does not exist at `v0.1.0`, so both it and this defect are entirely inside the unreleased cycle.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved numerical precision when calculating weighted log-mean-exp values, particularly for values near zero.
  * Ensured identical input values return exact results.
  * Improved handling of extreme, infinite, and mixed-value inputs.
  * Corrected lower-tail complementary log-log gradient calculations for heterogeneous profiles and extreme predictors.

* **Tests**
  * Added coverage for edge cases and gradient consistency.

### CODE REVIEW

**Comment by choxos** [comment]
@codex review

File: R/link_functions.R [hunk 1]
```diff
@@ -212,6 +212,35 @@ inverse_link <- function(x, link = c("identity", "log", "logit", "probit", "clog
...
+  # `log1p(sum(w * expm1(x)) / sum(w))` is the same mean with nothing to
+  # cancel: every `expm1(x_i)` keeps full relative accuracy for tiny `x_i`,
+  # the terms share a sign so the sum does too, and `log1p` inverts it. It
+  # is used only where it is also the safe form, `m_x` in `(-1, 0]`, since
+  # `expm1` of a large positive value overflows and a mean that underflows
+  # to zero would come back as `log1p(-1)`. Outside that range the shift is
+  # the accurate one: its error is absolute and of order `eps`, which only
+  # matters when the answer itself is near zero.
+  if (m_x <= 0 && m_x > -1) {
+    mean_expm1 <- sum(weights * expm1(x)) / sum(weights)
```
[the remark this answers is not in the document]
Comment (choxos) [reply R/link_functions.R]: Confirmed and fixed in 152f390, with a second failure mode the finding did not name.

Reproduced as stated: `weights = c(1e308, 1e308)` overflows `sum(weights)` to `Inf` and returns 0 where the mean is -7.898e-18. The numerator overflows too, since in the `m_x` range `(-1, 0]` each term reaches `0.632 * 1e308`, so two of them leave `-Inf / Inf` and a `NaN` that `if (mean_expm1 > -1)` errors on rather than returns.

The other end fails the same way and is not an overflow: `weights = c(1e-320, 1e-320)` underflows the NUMERATOR to 0 while the denominator stays representable, so it also returns 0 for that same -7.898e-18. A fix that only guarded the overflow would have left that one.

Normalizing by `max(weights)` handles both, and it is what the shifted form below was getting for free by working in logs. Every weight reaching that point is finite and positive, the zeros having been dropped earlier in the function, so the largest is a safe divisor and the ratio is unchanged by it. The scaled weights then lie in `(0, 1]` and each numerator term in `(-1, 1)`, so neither sum can leave double precision for any input length.

Measured after the change, against the algebraic reference `sum(scaled * x) / sum(scaled)`:

| weights | before | after |
|---|---:|---:|
| `c(1e308, 1e308)` | 0 | -7.898289e-18 |
| `c(1e-320, 1e-320)` | 0 | -7.898289e-18 |
| `c(2e307, 6e307)` | correct | correct |

The wide path is asserted unchanged at the same weights, since it never had the problem. Full suite 4719 pass, 0 fail, 0 error.

**Comment by choxos** [comment]
@codex review

File: R/link_functions.R (hunk 1 above)
[the remark this answers is not in the document]
Comment (choxos) [reply R/link_functions.R]: Already fixed, in 152f390, which landed after the commit this review read (`3d92b75`). Same defect CodeRabbit raised independently.

Reproduced as described: `weights = c(1e308, 1e308)` overflows `sum(weights)` to `Inf` and returns 0 where the mean is -7.898e-18. The numerator overflows too, since in this range each term reaches `0.632 * 1e308`, so two of them give `-Inf / Inf` and a `NaN` the guard errored on.

One addition: the other end fails the same way and is not an overflow. `weights = c(1e-320, 1e-320)` underflows the NUMERATOR to zero while the denominator stays representable, and also returns 0 for that same -7.898e-18. Scaling by `max(weights)` covers both, which is what the shifted form gets for free from working in logs. Every weight reaching here is finite and positive, the zeros having been dropped earlier, so the largest is a safe divisor.

File: R/link_functions.R [hunk 2]
```diff
@@ -212,6 +212,35 @@ inverse_link <- function(x, link = c("identity", "log", "logit", "probit", "clog
...
+    if (mean_expm1 > -1) return(log1p(mean_expm1))
```
[the remark this answers is not in the document]
Comment (choxos) [reply R/link_functions.R]: Confirmed and fixed in acca951. This one was live on the head, and it is the better of the two findings.

Reproduced: `x = c(0, -100)` with `weights = c(3, 1e16)` averages to -0.99999999999999978, whose `log1p` is **-36.043653**, against a mean of `log(3 / (3 + 1e16))` = **-35.742749**. That matches the reference you state. The maximum is zero, so the gate on the maximum cannot catch it: it is the MEAN that is far from one, dominated by a hugely weighted very negative value.

The diagnosis I take from it is that `mean_expm1 > -1` was gated on where the arithmetic BREAKS rather than where it stops being ACCURATE, and those are different points. Rather than reject a tuned neighborhood of -1, the branch is now gated on `mean_expm1 > -0.5`:

- this form exists for means near one, which is the regime where the shifted form loses the correction that is the whole answer;
- past -0.5 the shifted form is accurate anyway, since its error is absolute and of order eps against an answer of at least `log(0.5)`, a relative 3e-16.

So the two forms divide on which one is accurate, with no threshold that needs justifying against rounding.

The accuracy sweep gained 50 cases in exactly this regime, a maximum near zero with the mean dominated by a hugely weighted very negative value. Against a 60-digit reference over 170 designs: none exceeds a relative 1e-13, worst case 5.9e-16. Full suite 4722 pass, 0 fail, 0 error.

**Comment by choxos** [comment]
@codex review
